File: backend/agent_orchestrator.py
"""
Agent orchestrator — integrates deepagents as the AI reasoning layer.
Manages conversation context and routes between vision/audio/response.
"""

import asyncio
import logging

logger = logging.getLogger(__name__)

_agent = None
_agent_init_lock = asyncio.Lock()


def _get_openai_client():
    """Lazy import to avoid circular dependency with main.py."""
    from main import state
    return state.openai_client


async def get_agent():
    """
    Lazy-initialize and return the deepagents agent.
    Falls back to None (simple fallback) in demo mode or on failure.
    """
    global _agent

    if _agent is not None:
        return _agent

    async with _agent_init_lock:
        if _agent is not None:
            return _agent

        from main import state

        # In demo mode, skip deepagents entirely
        if not state.api_key_available:
            logger.info("Demo mode: deepagents disabled, using fallback processor")
            _agent = None
            return _agent

        # Try to initialize deepagents with a timeout
        try:
            logger.info("Initializing deepagents agent...")

            def _init_agent():
                from deepagents import create_deep_agent
                return create_deep_agent(
                    model="openai:gpt-4o",
                    system_prompt="""你是 AI 视觉对话助手，运行在用户手机上。
你可以实时看到用户摄像头画面、听到用户说的话。

## 核心原则
1. 回答尽量简洁（1-3句话），适合语音播报，不要用 markdown 格式
2. 如果用户没说话但画面有变化，主动描述画面
3. 如果用户问了关于画面内容的问题，结合画面信息回答
4. 使用中文回答，除非用户用其他语言提问

## 工作方式
- 你收到的消息格式是：画面描述 + 用户语音文本
- 基于这两部分信息综合分析并回应""",
                    tools=[analyze_visual_context, get_conversation_summary],
                )

            _agent = await asyncio.wait_for(
                asyncio.get_event_loop().run_in_executor(None, _init_agent),
                timeout=30.0,
            )
            logger.info("deepagents agent initialized successfully")
        except asyncio.TimeoutError:
            logger.warning("deepagents init timed out after 30s, using fallback")
            _agent = None
        except Exception as e:
            logger.warning("deepagents init failed: %s, using fallback", e)
            _agent = None

        return _agent


async def _fallback_process(user_text: str, frame_description: str | None = None) -> str:
    """Simple fallback when deepagents is unavailable."""
    from main import state

    # If we have an API key, do a direct OpenAI call
    if state.api_key_available and state.openai_client:
        try:
            messages = []
            if frame_description:
                messages.append({
                    "role": "user",
                    "content": f"画面内容: {frame_description}"
                })
            messages.append({
                "role": "user",
                "content": user_text
            })

            response = await state.openai_client.chat.completions.create(
                model="gpt-4o-mini",
                messages=[{
                    "role": "system",
                    "content": "你是视觉对话助手。回答简洁（1-3句），不要用markdown，使用中文。"
                }] + messages,
                max_tokens=300,
            )
            return response.choices[0].message.content.strip()
        except Exception as e:
            logger.warning("Fallback OpenAI call failed: %s", e)

    # Last resort: echo reply in demo mode
    if frame_description:
        return f"我看到：{frame_description[:80]}... 你说：{user_text}（演示模式 — 未配置 API Key）"
    return f"你说：{user_text}（演示模式 — 请设置 OPENAI_API_KEY 使用完整功能）"


async def analyze_visual_context(session_id: str | None = None) -> str:
    """
    Tool: retrieve the latest camera frame description for the session.
    Called by deepagents when it needs visual context.
    """
    try:
        from cost_controller import frame_controller
        if session_id:
            desc = frame_controller.get_latest_description(session_id)
            if desc:
                return f"当前画面: {desc}"
        return "当前画面: 暂无画面信息（等待新帧）"
    except Exception as e:
        logger.warning("analyze_visual_context error: %s", e)
        return "当前画面: 获取失败"


async def get_conversation_summary(session_id: str | None = None) -> str:
    """
    Tool: return a summary of the recent conversation history.
    Used for context management when token budget is tight.
    """
    try:
        from cost_controller import conversation_budget
        if session_id:
            messages = conversation_budget.get_recent_context(session_id, 5)
            if messages:
                parts = [f"{m['role']}: {m['content']}" for m in messages]
                return "最近对话:\n" + "\n".join(parts)
        return "暂无对话历史"
    except Exception as e:
        logger.warning("get_conversation_summary error: %s", e)
        return "获取历史失败"


async def process_user_input(
    session_id: str,
    user_text: str,
    frame_description: str | None = None,
) -> str:
    """
    Main conversation loop:
    1. Get/init the agent
    2. Build context (visual + speech)
    3. Invoke agent with context
    4. Track token usage
    5. Return reply text
    """
    agent = await get_agent()

    # Build the combined prompt with visual + audio context
    context_parts = []
    if frame_description:
        context_parts.append(f"[画面]: {frame_description}")
    context_parts.append(f"[用户]: {user_text}")

    prompt = "\n".join(context_parts)

    logger.info("Agent input (session=%s): %.150s", session_id, prompt)

    from cost_controller import conversation_budget

    try:
        if agent is None:
            # Fallback when deepagents is unavailable
            reply = await _fallback_process(user_text, frame_description)
        else:
            # Invoke the agent using deepagents (sync call in executor)
            def _invoke():
                result = agent.invoke({"messages": [{"role": "user", "content": prompt}]})
                messages = result.get("messages", [])
                for msg in messages:
                    if msg.get("role") == "assistant" and msg.get("content"):
                        return msg["content"]
                if messages:
                    return messages[-1].get("content", "")
                return ""

            reply = await asyncio.wait_for(
                asyncio.get_event_loop().run_in_executor(None, _invoke),
                timeout=60.0,
            )

        if not reply:
            return "嗯，我在听。"

        # Rough token estimate for cost tracking
        estimated_input_tokens = len(prompt) // 2
        estimated_output_tokens = len(reply) // 2

        conversation_budget.add_message(session_id, "user", user_text, estimated_input_tokens)
        conversation_budget.add_message(session_id, "assistant", reply, estimated_output_tokens)

        # Trigger summarization if budget exceeded
        if agent is not None and conversation_budget.needs_summary(session_id):
            try:
                def _summarize():
                    summary_response = agent.invoke({
                        "messages": [{
                            "role": "user",
                            "content": f"请用一句话总结以上对话，保留关键信息。原始对话内容：{reply}"
                        }]
                    })
                    return summary_response.get("messages", [{}])[-1].get("content", reply)

                summary = await asyncio.wait_for(
                    asyncio.get_event_loop().run_in_executor(None, _summarize),
                    timeout=30.0,
                )
                conversation_budget.replace_with_summary(
                    session_id, summary, len(summary) // 2
                )
                logger.info("Conversation summarized for session=%s", session_id)
            except Exception as e:
                logger.warning("Summarization failed: %s", e)

        logger.info("Agent reply (session=%s): %.100s", session_id, reply)
        return reply

    except asyncio.TimeoutError:
        logger.error("Agent invocation timed out (session=%s)", session_id)
        return "抱歉，处理超时，请再试一次。"
    except Exception as e:
        logger.error("Agent invocation failed: %s", e, exc_info=True)
        return f"抱歉，处理出错: {str(e)[:100]}"

File: agent_orchestrator.js
/**
 * Agent 编排 — 多模态上下文融合 + 对话管理
 */

const { ConversationBudget } = require('./cost_controller');

// 配置
const CHAT_MODEL = process.env.CHAT_MODEL || 'step-3.7-flash';
const MAX_TOKENS = parseInt(process.env.MAX_TOKENS || '4096', 10);

// 每个会话的对话历史和预算
const sessions = new Map();

/**
 * 系统提示词
 */
const SYSTEM_PROMPT = `你是一个 AI 视觉对话助手。你能够通过摄像头看到用户，通过麦克风听到用户说话。
请基于画面内容和用户的语音输入，给出自然、有帮助的回应。

要求：
- 用中文回复，口语化、自然，像朋友聊天一样
- 回复长度适中（2-4句话），不要长篇大论
- 如果画面中有值得注意的事物，可以主动提及
- 保持友好、积极的语气
- 如果用户问问题，直接回答
- 可以适当使用语气词让对话更自然`;

/**
 * 获取或创建会话
 * @param {string} sessionId
 * @returns {{ history: Array, budget: ConversationBudget }}
 */
function getSession(sessionId) {
  if (!sessions.has(sessionId)) {
    sessions.set(sessionId, {
      history: [{ role: 'system', content: SYSTEM_PROMPT }],
      budget: new ConversationBudget(MAX_TOKENS),
      frameDescription: '尚未获取画面',
    });
  }
  return sessions.get(sessionId);
}

/**
 * 更新当前帧描述
 * @param {string} sessionId
 * @param {string} description
 */
function updateFrameDescription(sessionId, description) {
  const session = getSession(sessionId);
  session.frameDescription = description;
}

/**
 * 对话推理 — 融合画面 + 语音
 * @param {object} client - OpenAI 客户端
 * @param {string} sessionId - 会话 ID
 * @param {string} userText - 用户说的话
 * @param {boolean} apiAvailable - API 是否可用
 * @returns {Promise<string>} AI 回复文本
 */
async function processUserInput(client, sessionId, userText, apiAvailable) {
  const session = getSession(sessionId);
  const { history, budget, frameDescription } = session;

  // Token 预算检查：超限时自动摘要
  if (budget.needsSummarization()) {
    await summarizeHistory(client, sessionId, apiAvailable);
  }

  // Demo 模式：mock 回复
  if (!apiAvailable) {
    const mockReplies = [
      `我看到了画面。${frameDescription}你想聊些什么呢？`,
      '你好！我在这里听着呢。虽然现在是演示模式，但我们可以继续对话。',
      '这是一个有趣的观察。配置好 API Key 后，我能给你更智能的回复。',
    ];
    const reply = mockReplies[Math.floor(Math.random() * mockReplies.length)];
    budget.addTokens(ConversationBudget.estimateTokens(reply));
    history.push({ role: 'assistant', content: reply });
    return reply;
  }

  // 构建用户消息（融合画面描述 + 语音文本）
  const userMessage = frameDescription && frameDescription !== '尚未获取画面'
    ? `[当前画面] ${frameDescription}\n[用户说] ${userText}`
    : `[用户说] ${userText}`;

  history.push({ role: 'user', content: userMessage });

  try {
    const response = await client.chat.completions.create({
      model: CHAT_MODEL,
      messages: history,
      max_tokens: 500,
      temperature: 0.8,
    });

    const reply = response.choices[0]?.message?.content || '抱歉，我没有理解你说的话。';

    // 记录 Token 消耗
    if (response.usage) {
      budget.addTokens(response.usage.total_tokens);
    } else {
      budget.addTokens(ConversationBudget.estimateTokens(userMessage + reply));
    }

    // 保持对话历史
    history.push({ role: 'assistant', content: reply });

    // 控制历史长度（保留最近 20 条消息 + system prompt）
    if (history.length > 21) {
      // 保留 system prompt + 最近 20 条
      const systemMsg = history[0];
      session.history = [systemMsg, ...history.slice(-20)];
    }

    return reply;
  } catch (err) {
    console.error('[Agent] 推理失败:', err.message);
    const fallbackReply = '抱歉，我暂时无法处理你的请求，请稍后再试。';
    history.push({ role: 'assistant', content: fallbackReply });
    return fallbackReply;
  }
}

/**
 * 纯文本对话（无画面上下文）
 * @param {object} client - OpenAI 客户端
 * @param {string} sessionId
 * @param {string} userText
 * @param {boolean} apiAvailable
 * @returns {Promise<string>}
 */
async function chatText(client, sessionId, userText, apiAvailable) {
  return processUserInput(client, sessionId, userText, apiAvailable);
}

/**
 * 摘要对话历史（Token 预算超限时）
 * @param {object} client
 * @param {string} sessionId
 * @param {boolean} apiAvailable
 */
async function summarizeHistory(client, sessionId, apiAvailable) {
  const session = getSession(sessionId);
  const { history, budget } = session;

  if (history.length <= 3 || !apiAvailable) {
    // 简单截断：保留 system prompt + 最近 6 条
    session.history = [history[0], ...history.slice(-6)];
    budget.reset();
    // 估算重置后的 tokens
    const remaining = history[0].content + session.history.slice(1).map(m => m.content).join('');
    budget.addTokens(ConversationBudget.estimateTokens(remaining));
    return;
  }

  try {
    const historyCopy = history.slice(1); // 去掉 system prompt
    const summaryResponse = await client.chat.completions.create({
      model: CHAT_MODEL,
      messages: [
        {
          role: 'system',
          content: '请用1-2句话总结以下对话的核心内容，保留关键信息。',
        },
        {
          role: 'user',
          content: historyCopy.map(m => `[${m.role}]: ${m.content}`).join('\n'),
        },
      ],
      max_tokens: 200,
      temperature: 0.3,
    });

    const summary = summaryResponse.choices[0]?.message?.content || '对话历史已压缩。';

    // 重建历史：system prompt + 摘要 + 最近 4 条
    session.history = [
      history[0],
      { role: 'system', content: `[对话摘要] ${summary}` },
      ...history.slice(-4),
    ];
    budget.reset();
    const totalText = session.history.map(m => m.content).join('');
    budget.addTokens(ConversationBudget.estimateTokens(totalText));

    console.log(`[Agent] 对话已压缩，摘要: ${summary.substring(0, 80)}...`);
  } catch (err) {
    console.warn('[Agent] 摘要失败，使用截断:', err.message);
    session.history = [history[0], ...history.slice(-6)];
    budget.reset();
  }
}

/**
 * 清理会话
 * @param {string} sessionId
 */
function cleanupSession(sessionId) {
  sessions.delete(sessionId);
}

module.exports = {
  getSession,
  updateFrameDescription,
  processUserInput,
  chatText,
  summarizeHistory,
  cleanupSession,
};
